spring-shell-core/src/main/java/org/springframework/shell/command/CommandCatalog.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.shell.command;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.springframework.shell.context.InteractionMode;
+import org.springframework.shell.context.ShellContext;
+import org.springframework.util.StringUtils;
+
+/**
+ * Interface defining contract to handle existing {@link CommandRegistration}s.
+ *
+ * @author Janne Valkealahti
+ */
+public interface CommandCatalog {
+
+	/**
+	 * Register a {@link CommandRegistration}.
+	 *
+	 * @param registration the command registration
+	 */
+	void register(CommandRegistration... registration);
+
+	/**
+	 * Unregister a {@link CommandRegistration}.
+	 *
+	 * @param registration the command registration
+	 */
+	void unregister(CommandRegistration... registration);
+
+	/**
+	 * Unregister a {@link CommandRegistration} by its command name.
+	 *
+	 * @param commandName the command name
+	 */
+	void unregister(String... commandName);
+
+	/**
+	 * Gets all {@link CommandRegistration}s mapped with their names.
```

Comment:
I know this is a high level review - but I would eventually elaborate a bit more here as the default impl resolves dynamic commands as well as filter based on interactive/non-interactive. 

Also, I think its important to mention the map is a copy and any updates made to it will be ignored.

Reply:
This is a bit difficult design part of this interface. As you already commented an actual command name is and its use is spanning `CommandCatalog` and `CommandRegistration`. I originally had this just returning a list of registrations but then you need to calculate names out from registrations which felt weird.